### ISSUE TITLE
Support globbing in `banned-api`

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#2660]
E.g., `top` should match `import top`, but not `import top.internal`, whereas `top.*` should match both (I think? We'll do whatever `flake8-tidy-imports` does).

See: #2656.

**Comment by not-my-profile** [comment]
No ... I don't think that's a good idea because `import top.internal` also imports `top`.

**Comment by charliermarsh** [comment]
Yes good point! I agree that we shouldn't implement the submodule semantics described above. It was a lazy example.

The question here is really: what does `flake8-tidy-imports` do? And do we want to support _that_? (E.g., do we want to support globbing like `example.*.truck`, as in `flake8-tidy-imports`?)

(I don't really want to support it, and I won't be working on it any time soon, but that's the intent of the issue.)

**Comment by seamuswn** [comment]
We have a Python monorepo in my org, where we have developed shared abstractions built under `src.core`. Ideally we would like to prevent importing private / internal packages within that core package. Specifically restricting subpackages at any level  within `src.core` named 'internal' (i.e. `src.core.*.internal`), or any subpackages where the package name starts with `_` (i.e. `src.core._*`/`src.core.*._*`)

**Comment by Avasam** [comment]
If globbing (or at least submodules) support is added to https://docs.astral.sh/ruff/rules/banned-api/ (TID251), could it also be added to https://docs.astral.sh/ruff/rules/banned-module-level-imports/ (TID253) ? I'd like to configure it like such:
```py
[lint.flake8-tidy-imports]
banned-module-level-imports = [
	# platform-specific imports
	"android",
	"jnius",
	"platformdirs.android", # still allowing top-level platformdirs
	"platformdirs.linux" ,# still allowing top-level platformdirs
	"platformdirs.macos", # still allowing top-level platformdirs
	"platformdirs.windows", # still allowing top-level platformdirs
	"winreg",
]
```
And at this point, https://docs.astral.sh/ruff/rules/banned-import-from/ (ICN003) as well.

**Comment by rh-sp** [comment]
To expand on @ashrub-holvi's take, I would like to limit import to canonical locations in a Django app. So for example, models should always be imported from `my_app.models`, never `my_app.models.foo` or `my_app.models.bar`. A more pressing version of this, a colleague of mine has identified an issue with Django signals. Not sure how much of this is our setup vs. Django conventions, but at least in our case, signals are only registered if they're imported in `my_app.signals`. If we allow importing a signal from `my_app.signals.foo`, it will not work if it's not also imported by `my_app.signals`'s `__init__.py`.

So Ideally I would like to have some setting like:

```toml
[lint.something]
banned-something = [
    "*.signals.**",
    "*.models.**"
]
```

Which would lead to the following:

```python
from my_app.models import Product  # allowed
from my_app.models.product import Product  # banned
from my_app.models.product.simple import SimpleProduct  # banned (** matches one or more components)
from something_else.foo.models.product import Product  # allowed (* matches only single component)
```

The last line being allowed isn't required for us, but it would be nice to be able to make this distinction.
